packages/api/endpoints/collections.js

```diff
@@ -216,24 +226,45 @@ async function put(req, res) {
  * @returns {Promise<Object>} the promise of express response object
  */
 async function del(req, res) {
+  const {
+    collectionsModel = new models.Collection(),
+    collectionPgModel = new CollectionPgModel(),
+    knex = await getKnexClient(),
+    esClient = await Search.es(),
+  } = req.testContext || {};
+
   const { name, version } = req.params;
-  const collectionsModel = new models.Collection();
 
-  const knex = await getKnexClient({ env: process.env });
+  let existingCollection;
   try {
-    await knex.transaction(async (trx) => {
-      await trx(tableNames.collections).where({ name, version }).del();
-      await collectionsModel.delete({ name, version });
-      if (inTestMode()) {
+    existingCollection = await collectionsModel.get({ name, version });
+  } catch (error) {
+    if (!(error instanceof RecordDoesNotExist)) {
+      throw error;
+    }
+  }
+
+  try {
+    try {
+      await knex.transaction(async (trx) => {
+        await collectionPgModel.delete(trx, { name, version });
+        await collectionsModel.delete({ name, version });
         const collectionId = constructCollectionId(name, version);
-        const esClient = await Search.es(process.env.ES_HOST);
         await esClient.delete({
           id: collectionId,
           index: process.env.ES_INDEX,
           type: 'collection',
+          refresh: inTestMode(),
         }, { ignore: [404] });
+      });
+    } catch (innerError) {
+      // Delete is idempotent, so there may not be a DynamoDB
+      // record to recreate
+      if (existingCollection) {
+        await collectionsModel.create(existingCollection);
```

Comment:
In other PRs we talked about not needing a similar catch on `delete` but I think it's necessary here. We have the `esClient.delete` coming after both model deletes so if _that_ fails, we might have already deleted the Dynamo collection and will need to re-create.

Reply:
Exactly